packages/jira-adapter/src/filters/fields/context_options.ts

```diff
@@ -95,30 +97,102 @@ type UpdateContextOptionsParams = {
   addedOptions: Value[]
   modifiedOptions: Value[]
   removedOptions: Value[]
-  client: clientUtils.HTTPWriteClientInterface
+  client: JiraClient
   baseUrl: string
   contextChange: ModificationChange<InstanceElement> | AdditionChange<InstanceElement>
+  paginator: clientUtils.Paginator | undefined
+  isCascade: boolean
+  numberOfAlreadyAddedOptions: number
+}
+
+type Option = {
+  id: string
+  value: string
+  optionId?: string
+}
+
+const EXPECTED_OPTION_SCHEMA = Joi.object({
+  id: Joi.string().required(),
+  value: Joi.string().required(),
+}).unknown(true)
+
+const isOption = createSchemeGuard<Option>(EXPECTED_OPTION_SCHEMA, 'Received invalid response from private API')
+
+const proccessContextOptionsPrivateApiResponse = (
+  resp: Option[],
+  contextChange: AdditionChange<InstanceElement> | ModificationChange<InstanceElement>,
+): void => {
+  const idToOption = _.keyBy(contextChange.data.after.value.options, option => option.id)
+  const optionsMap = _(contextChange.data.after.value.options).values()
+    .keyBy(option => naclCase(option.value)).value()
+  const optionIds = new Set(Object.keys(idToOption))
+  const respToProccess = resp.filter(newOption => !optionIds.has(newOption.id))
+  respToProccess.forEach(newOption => {
+    if (newOption.optionId !== undefined) {
+      idToOption[newOption.optionId]
+        .cascadingOptions[naclCase(newOption.value)].id = newOption.id
+    } else {
+      optionsMap[naclCase(newOption.value)].id = newOption.id
+    }
+  })
+}
+const getAllOptionPaginator = async (paginator: clientUtils.Paginator, baseUrl: string)
+  : Promise<Option[]> => {
+  const paginationArgs: clientUtils.ClientGetWithPaginationParams = {
+    url: baseUrl,
+    paginationField: 'startAt',
+    queryParams: {
+      maxResults: '1000',
+    },
+    pageSizeArgName: 'maxResults',
+  }
+
+  const optionsPromise = toArrayAsync(paginator(
+    paginationArgs,
+    page => makeArray(page) as clientUtils.ResponseValue[]
+  ))
+  const values = (await optionsPromise).flatMap(lst => lst[0].values).filter(isOption)
```

Comment:
I think you are supposed to do 
```suggestion
  const options = await toArrayAsync(paginator(
    paginationArgs,
    page => makeArray(page.values) as clientUtils.ResponseValue[]
  ))
  const values = options.filter(isOption)
```

Reply:
options is still list of lists, so I think I need to have `flat()` as well. 
`const values = options.flat().filter(isOption)`